### ISSUE TITLE
States function incorrectly when multiple layers are used.

### ISSUE THREAD

**Issue by justonia** [DigitalMachinist/unity-utilities#1]
With the use of layers in Mechanim it is conceptually equivalent to being in multiple states at once. The current implementation of the state machine utilities will incorrectly transition out of states when transitions occur in different layers. 

I addressed this by adding an 'int layerIndex' field to StateMachine, adding multiple StateMachines to my Animator's game object, and then when a State boots up in OnStateEnter and searches for a StateMachine, it makes sure that it picks the one with the matching layer.

Additionally, it might be worth noting in a "Limitations" section that the State class will not function at all correctly if added to a Sub-State Machine node. The reason for this is that the OnStateEnter/Exit callbacks are called for every single state and transition inside the sub-state machine. I address this by only ever putting State classes on non-Sub-State nodes.

**Comment by DigitalMachinist** [comment]
Yeah, you're right. Thank you for pointing this out. I should really clarify this as a limitation. I noted that layers would probably cause problems a little while ago but I never really used them myself so I wasn't worrying about it.

This is probably the first I've had any evidence that anyone else was using this library for anything. Up until now it's been nothing more than an personal toolkit that's kinda helpful to me.

I hope it's helping you out with whatever you're doing. I'll have to update my docs to express this issue and maybe take some time in the future to see if there's a good way that I can provide a good abstraction for layering that leads to more helpful behaviour.

I guess one major item would be to find a way to think about layers as separate, where transition events from a layer don't fire events from those on some other layer. Since you developed a workaround for getting layers to work for you, do you have any thoughts on how this might be most usefully handled? How are you using layers here and how might state management be thought about better for what you're trying to do?

**Comment by justonia** [comment]
Currently one of my character controllers has two layers, "Base Layer" and "Weapons". Weapons is an overlay layer and is allowed to control the state of the character's weaponry usage, such as aiming, shooting, etc. The base layer manages mostly the lower body's locomotion, idle poses, etc. The weapon overlay layer can take over arms to have the character be firing while being in most any base state.

Since weaponry is separate from the base layer controls, it makes sense for me to have two entirely different state machine instances controlling their primary function.

The only problem I haven't solved is the callback invocation. Currently each instance of the StateMachine is subscribing to all States on every layer. Obviously this is bad, so what I am thinking of doing is deferring callback registration on the StateMachine until the first time OnStateEnter is called so I have access to the layerIndex. Given that the the callback never would've been triggered from that state until it has run once, it seems like a reasonable approach.

Oh. Also I found one other issue that I fixed in my version. If you transition out of a Mechanim node into another node that does not have a State behaviour on it, by the time you reach the next node with a State on it you have lost the context of what the last State you came from is, and worse, OnControlEnter is never called on the new State. I came up with the concept of a "deferred exit" to handle this.

I've made a number of personal changes to these files, so a pull request would not really be appropriate. Here is a gist you can reference though:

https://gist.github.com/justonia/d500e5ce0f8d28cb6713a78491587d92

Thanks for your work on this utility. I was beating my head against a wall to get reliable ordering of callbacks between states while undergoing transitions. I'm sure I'll run into some weird edge case with interruptions on transitions, but maybe Unity will play nice and send the correct events.

**Comment by justonia** [comment]
I also broke StateMachine into two files, the StateMachine<T> so I can have different types of machines for each layer residing on the same game object.

**Comment by DigitalMachinist** [comment]
Thanks for the gist. It'll help with figuring out what makes the most sense to do in future.

I can definitely understand this. Lots of games have a few context specific controls that change subsets of the controls. It would make handling controls of that kind a lot easier and less repetitive (and possibly less buggy) if you only had to think about what controls are changing for each weapon instead of resetting the entire control context for each and every one of them.

> The only problem I haven't solved is the callback invocation. Currently each instance of the StateMachine is subscribing to all States on every layer.

Yeah, I found Unity's Animator API to be pretty sparse when I was working on this initially just to do what this utility already does. Admittedly I wasn't thinking about layers at the time, but now that we're considering them I'm not entirely sure the Animator's public API will give me the information I need to make a clean distinction between layers, so it might have to be something a little bit hackier like the layerIndex thing you're doing to work around this. Hopefully I can find a good way to encapsulate this away and make it easy to use.

> If you transition out of a Mechanim node into another node that does not have a State behaviour on it, by the time you reach the next node with a State on it you have lost the context of what the last State you came from is, and worse, OnControlEnter is never called on the new State. I came up with the concept of a "deferred exit" to handle this.

Oh wow, yeah that's a pretty serious issue. When I work on this I'll probably bring your solution across since it seems to make sense that there be some kind of default handling to provide continuity where no State has been defined between a couple of other States.

**Comment by DigitalMachinist** [comment]
I decided to work on this this evening and I think I have a solid plan to make layers work.

I was curious about one thing in your gist, though. Is there a specific reason that you chose to hide the events in State from the inspector? 

``` csharp
[NonSerialized] public StateMachineEvent StateEnter = new StateMachineEvent();
[NonSerialized] public StateMachineEvent StateUpdate = new StateMachineEvent();
[NonSerialized] public StateMachineEvent StateExit = new StateMachineEvent();
[NonSerialized] public StateMachineEvent ControlEnter = new StateMachineEvent();
[NonSerialized] public StateMachineEvent ControlUpdate = new StateMachineEvent();
[NonSerialized] public StateMachineEvent ControlExit = new StateMachineEvent();
```

I'm guessing that it reduces confusion about which events you should register to from elsewhere and maybe generally declutters the inspector, but I was curious if you had a deeper reason for this change.

**Comment by DigitalMachinist** [comment]
If you're able, I would give this new build a shot, but I realize your fork of the project might be different enough that you can't just swap this in so easily.

I've added in a list of AnimatorLayer objects to each AnimatorStateMachine (a few naming tweaks happened). Each layer has its own context so layers don't trample each other's shared state (since they don't share it anymore).

I also made sure that, by default, if no AnimatorState is attached to a Mechanim state, it will be handled approximately like your DeferredStateExit, but with a few modifications. So now there's no reason to worry about things breaking down if you forget a state somewhere anymore.

It's probably still got a few bugs, but I'll try to make up a couple demos with of the new features to iron out the wrinkles.

Thanks again for your input.
